src/app.ts

```diff
@@ -12,15 +12,74 @@ interface TemperatureSummary {
   low: number
   average: number
 }
+/* dictionary of dictionarys for the summarys, i use a counter for  */
+interface IndexedSumary {
+  [city: string]: {
+    [time: string]: {
+      summary: TemperatureSummary
+      count: number
+    }
+  }
+}
+const temperatures: IndexedSumary = {}
 
 export function processReadings(readings: TemperatureReading[]): void {
-  // add here your code
+  //add here your code
+  /*we iterate over all the readings, if there isn't a valid entry for a key, it creates it
+  each read updates the values if it is necesary
+  
+  i used for of instead of reduce because its faster as far as I have been able to research
+  and 1 loop does less comparations than 3*/
+  for (const read of readings) {
+    const readTimeToString: string = read.time.toString()
+    if (temperatures[read.city] == undefined) {
+      temperatures[read.city] = {}
+    }
+    if (temperatures[read.city][readTimeToString] == undefined) {
+      temperatures[read.city][readTimeToString] = {
+        summary: {
+          first: read.temperature, // if the key doesn't had an assigned value then it's the first measure
+          last: read.temperature,
+          high: read.temperature,
+          low: read.temperature,
+          average: read.temperature,
+        },
+        count: 1,
+      }
+    } else {
+      const tempAux: TemperatureSummary =
+        temperatures[read.city][readTimeToString].summary
+      //each measure would be the last if there is no other new
+      tempAux.last = read.temperature
+      //we update the higher measure if it's necesary
+      if (tempAux.high < read.temperature) tempAux.high = read.temperature
+      //we update the lower measure if it's necesary
+      else if (tempAux.low > read.temperature) tempAux.low = read.temperature
+      //we add up all the values for the average
+      tempAux.average += read.temperature
+      //we count the values for the average
+      temperatures[read.city][readTimeToString].count++
+    }
+  }
+  //we update the average knowing the sum of all measures of a day and the count of these
+  for (const keyCity in temperatures) {
+    for (const keyTime in temperatures[keyCity]) {
+      temperatures[keyCity][keyTime].summary.average /=
+        temperatures[keyCity][keyTime].count
+    }
+  }
```

Comment:
I think you can eliminate the need for this double for loop. Give it a few minutes of thought and see what you can come up with. If you get stuck let me know 😉

Reply:
Also, I think the average temp here means the average between the low and high. Or in otherwise the average temperature for the day

Reply:
I have some ideas but I don't know if they would be correct:
- Recalculate the average each time I enter a new entry (I don't like the idea very much because of the low precision of the float and that it could cause erroneous results in the long run. )
```
for (const read of readings) {
    ...
    if (temperatures[read.city][read.time.getTime()] != undefined){
      ...
      tempAux.average = ((tempAux.average*tempAux.count)+read.temperature)/(tempAux.count+1)
    }
  }
```

Reply:
- use a composed key instead 2 keys for my dictionary
use `dictionary ['Utah01/02/2010']` instead of `dictionary ['Utah']['01/02/2010']`

Reply:
- I'm going to review how to use iterators, something tells me that is the right way

Reply:
After think in that for a while, i think the best is to **calculate the average at the user's request**, and perhaps that could be extended to all properties of the summary, and calculate the summary of a day just if the user request it

Reply:
Nice! Thanks for putting so much thought into this. I've been thinking about this a bit as well and there are 2 paths. You've either taken or mentioned both. 1. you can run the calculation for the historic average an all the data so that it's available when you ask for it. This is the method you originally used. The other is what you're describing above where you calculate when you need it. I think the answer to the question depends on where this code lives and what kind of quantity you're dealing with on the data

If the server is doing the work I'd guess that either approach would work but that SQL has some super efficient way to calculate an average. Initially I would worry about calculating the average on request in the client code because client code (especially in react) tends to update and be run often so you'd likely be calculating it over and over again on every render if you're not careful

I was originally thinking about the daily average when I mentioned you don't need the double for loop. If we're doing historic average it would seem we do need to go back and calculate by doing something like this. Unless you go with the incremental average approach